packages/context/src/context.ts

```diff
@@ -797,21 +799,27 @@ export class Context extends EventEmitter {
   ): ValueOrPromise<ValueType | undefined> {
     const {key, propertyPath} = BindingKey.parseKeyWithPath(keyWithPath);
 
-    optionsOrSession = asResolutionOptions(optionsOrSession);
-
-    const binding = this.getBinding<ValueType>(key, optionsOrSession);
-    if (binding == null) return undefined;
+    const options = asResolutionOptions(optionsOrSession);
+
+    let binding: Binding<ValueType> | undefined;
+    try {
+      binding = this.getBinding<ValueType>(key, options);
+      if (binding == null) return undefined;
+    } catch (err) {
+      throw new ResolutionError(err, {
+        context: this,
+        binding: Binding.bind(key),
+        options,
+      });
+    }
 
-    const boundValue = binding.getValue(this, optionsOrSession);
+    const boundValue = binding.getValue(this, options);
     if (propertyPath === undefined || propertyPath === '') {
       return boundValue;
     }
-
-    if (isPromiseLike(boundValue)) {
-      return boundValue.then(v => getDeepProperty<ValueType>(v, propertyPath));
-    }
-
-    return getDeepProperty<ValueType>(boundValue, propertyPath);
+    return transformValueOrPromise(boundValue, v =>
+      getDeepProperty<ValueType>(v, propertyPath),
+    );
```

Comment:
Nice clean up, would you mind extracting it into a standalone PR to get it landed sooner + make future code archaeology easier?

Reply:
See #5329

Reply:
I am afraid I don't see how #5329 is related to his change? Did you perhaps make a typo in the issue/PR number?

Reply:
Sorry, it's #5593